File: Tree/310.Minimum-Height-Trees/310.Minimum Height Trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<pair<int, int>>& edges) 
    {
        if (n==1) return {0};
        if (n==2) return {0,1};
        
        unordered_map<int,int>Degree;
        unordered_map<int,vector<int>>Up;
        
        for (auto edge:edges)
        {
            Degree[edge.first]++;
            Degree[edge.second]++;
            Up[edge.second].push_back(edge.first);
            Up[edge.first].push_back(edge.second);
        }
        
        queue<int>q;
        for (auto a:outDegree)
            if (a.second==1) q.push(a.first);
        
        int count = 0;
        while (!q.empty())
        {
            int k = q.size();
            for (int i=0; i<k; i++)
            {
                int node = q.front();
                q.pop();
                count++;

                for (auto upNode: Up[node])
                {
                    Degree[upNode]--;
                    if (Degree[upNode]==1)
                        q.push(upNode);
                }
            }
                       
            if (count+q.size()==n)
            {
                vector<int>results;
                while (!q.empty())
                {
                    results.push_back(q.front());
                    q.pop();
                }
                return results;
            }
        }
    }
};

File: Tree/310.Minimum-Height-Trees/310.Minimum-Height-Trees.py
class Solution:
    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        if (n==1): return [0]
        if (n==2): return [0,1]
    
        degree = [0]*n
        Neighbor = [[] for i in range(n)];
        for edge in edges:
            degree[edge[0]]+=1
            degree[edge[1]]+=1
            Neighbor[edge[0]].append(edge[1])
            Neighbor[edge[1]].append(edge[0])
        
        q = collections.deque()
        for i in range(n):
            if degree[i]==1:
                q.append(i)
                
        count = 0        
        while (q):
            L = len(q)
            for i in range(L):
                x = q.popleft()
                count+=1
                for y in Neighbor[x]:
                    degree[y]-=1
                    if degree[y]==1:
                        q.append(y)
            if count+len(q)==n:
                result = []
                while (q): result.append(q.popleft())
                return result
        
